swift-sdk/Internal/IterableAPIInternal.swift

```diff
@@ -0,0 +1,819 @@
+                // we are ready
+                IterableAppIntegration.implementation?.performDefaultNotificationAction(remoteNotificationPayload)
+            } else {
+                // keywindow not set yet
+                DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
+                    IterableAppIntegration.implementation?.performDefaultNotificationAction(remoteNotificationPayload)
+                }
+            }
+        }
+    }
+
+    private func checkForDeferredDeeplink() {
+        guard config.checkForDeferredDeeplink else {
+            return
+        }
+        guard localStorage.ddlChecked == false else {
+            return
+        }
+        
+        guard let request = IterableRequestUtil.createPostRequest(forApiEndPoint: ITBConsts.linksEndpoint, path: ENDPOINT_DDL_MATCH, args: [ITBL_KEY_API_KEY : apiKey], body: DeviceInfo.createDeviceInfo()) else {
+            ITBError("Could not create request")
+            return
+        }
+        
+        NetworkHelper.sendRequest(request, usingSession: networkSession).observe {(result) in
+            switch result {
+            case .value(let json):
+                self.handleDDL(json: json)
+            case .error(let failureInfo):
+                if let errorMessage = failureInfo.errorMessage {
+                    ITBError(errorMessage)
+                }
+            }
+        }
+    }
+    
+    private func handleDDL(json: [AnyHashable : Any]) {
+        if let serverResponse = try? JSONDecoder().decode(ServerResponse.self, from: JSONSerialization.data(withJSONObject: json, options: [])),
+            serverResponse.isMatch,
+            let destinationUrlString = serverResponse.destinationUrl,
+            let action = IterableAction.actionOpenUrl(fromUrlString: destinationUrlString) {
+            let context = IterableActionContext(action: action, source: .universalLink)
+            
+            DispatchQueue.main.async {
+                IterableActionRunner.execute(action: action,
+                                             context: context,
+                                             urlHandler: IterableUtil.urlHandler(fromUrlDelegate: self.urlDelegate, inContext: context),
+                                             urlOpener: AppUrlOpener())
+            }
+        }
+        
+        localStorage.ddlChecked = true
+    }
+    
+    private func updateDeviceId() {
+        if localStorage.deviceId == nil {
+            localStorage.deviceId = IterableUtil.generateUUID()
+        }
+    }
+    
+    private func updateSDKVersion() {
+        if let lastVersion = localStorage.sdkVersion, lastVersion != IterableAPI.sdkVersion {
+            performUpgrade(lastVersion: lastVersion, newVersion: IterableAPI.sdkVersion)
+        } else {
+            localStorage.sdkVersion = IterableAPI.sdkVersion
+        }
+    }
+    
+    private func performUpgrade(lastVersion: String, newVersion: String) {
```

Comment:
`String` isn't very useful when determining which upgrades to perform; I think transforming the version (1.2.3) into an int (1002003) would let us specify upgrade paths. That way, you can do `if lastVersion < 1002003` to figure out whether you need a specific upgrade. Alternatively, there could be a method like `shouldUpgradeFrom(oldVersion: String, newVersion: String)` that would do String -> int conversions and comparison internally.

Reply:
We can add those later. I would rather design this properly when we actually do upgrades. We have the plumbing here

Reply:
Yeah that's not necessary for now.